packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -416,59 +317,125 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
+            nonfungiblePositionManager,
+            ControllerHelperDataType.closeUniLpParams({
                 tokenId: _params.tokenId,
-                recipient: address(this),
-                amount0Max: type(uint128).max,
-                amount1Max: type(uint128).max
-            })
+                liquidity: _params.liquidity,
+                liquidityPercentage: 1e18,
+                amount0Min: uint128(_params.amount0Min),
+                amount1Min: uint128(_params.amount1Min)
+            }),
+            isWethToken0
         );
 
-        if (wPowerPerpAmount > 0) {
+        if (_params.wPowerPerpAmountDesired > wPowerPerpAmountInLp) {
+            // if the new position target a higher wPowerPerp amount, swap WETH to reach the desired amount (WETH new position is lower than current WETH in LP)
+            _exactOutFlashSwap(
+                weth,
+                wPowerPerp,
+                IUniswapV3Pool(wPowerPerpPool).fee(),
+                _params.wPowerPerpAmountDesired.sub(wPowerPerpAmountInLp),
+                _params.limitPriceEthPerPowerPerp.mul(_params.wPowerPerpAmountDesired.sub(wPowerPerpAmountInLp)).div(
+                    1e18
+                ),
+                uint8(CALLBACK_SOURCE.SWAP_EXACTOUT_ETH_WPOWERPERP),
+                ""
+            );
+        } else if (_params.wPowerPerpAmountDesired < wPowerPerpAmountInLp) {
+            // if the new position target lower wPowerPerp amount, swap excess to WETH (position target higher WETH amount)
+            uint256 wPowerPerpExcess = wPowerPerpAmountInLp.sub(_params.wPowerPerpAmountDesired);
+
             _exactInFlashSwap(
                 wPowerPerp,
                 weth,
                 IUniswapV3Pool(wPowerPerpPool).fee(),
-                wPowerPerpAmount,
-                _params.limitPriceEthPerPowerPerp.mul(wPowerPerpAmount).div(1e18),
+                wPowerPerpExcess,
+                _params.limitPriceEthPerPowerPerp.mul(wPowerPerpExcess).div(1e18),
                 uint8(CALLBACK_SOURCE.SWAP_EXACTIN_WPOWERPERP_ETH),
                 ""
             );
         }
 
+        // mint new position
+        ControllerHelperUtil.lpWPowerPerpPool(
+            nonfungiblePositionManager,
+            wPowerPerpPool,
+            ControllerHelperDataType.LpWPowerPerpPool({
+                recipient: msg.sender,
+                ethAmount: _params.ethAmountToLp,
+                amount0Desired: (isWethToken0) ? _params.wethAmountDesired : _params.wPowerPerpAmountDesired,
+                amount1Desired: (isWethToken0) ? _params.wPowerPerpAmountDesired : _params.wethAmountDesired, 
+                amount0Min: _params.amount0DesiredMin,
+                amount1Min: _params.amount1DesiredMin,
+                lowerTick: _params.lowerTick,
+                upperTick: _params.upperTick
+            })
+        );
+
+        ControllerHelperUtil.checkLpMintExcess(controller, wPowerPerp, nonfungiblePositionManager, 0);
```

Comment:
why do we call this here? 

Isn't there a case where we have more wPowerPerp or weth that is not needed in the LP position that we need to sell back/refund?  Might be worth seeing what the swap and add function in uniswap repo does here.